Sources/PackageLoading/TargetSourcesBuilder.swift

```diff
@@ -83,12 +83,44 @@ public struct TargetSourcesBuilder {
             }
         }
         self.declaredSources = declaredSources?.spm_uniqueElements()
+        
+        self.excludedPaths.forEach { exclude in
+            validTargetPath(at: exclude, message: "Cannot exclude path: \(exclude) because it does not exist")
+        }
+        
+        self.declaredSources?.forEach { source in
+            validTargetPath(at: source, message: "Source: \(source) does not exist")
+        }
 
       #if DEBUG
         validateRules(self.rules)
       #endif
     }
 
+    @discardableResult
+    private func validTargetPath(at: AbsolutePath, message: String?) -> Bool {
+        guard let cwd = self.fs.currentWorkingDirectory else {
+            return false
+        }
+        
+        // Check if paths that are enumerated in targets: [] exist
+        guard self.fs.exists(at) else {
+            if let msg = message {
+                self.diags.emit(warning: msg)
+            }
+            return false
+        }
+
+        // Excludes, Sources, and Resources should be found at the root of the package and or
+        // its subdirectories
+        guard at.pathString.hasPrefix(cwd.pathString) else {
+            self.diags.emit(warning: "\(at) should be found within the root of \(cwd)")
```

Comment:
should this case also include the user's message? it seems strange to me that in some case we take a "message" argument but in some cases dont use it. this could say "\(message). \(at) should be found within the root of \(cwd)""

Reply:
or maybe the right way to design this API is to make it into

`private func validTargetPath(at: AbsolutePath) ->[Error] {`

and then return all the errors if any found, and the call site can do something like

```
self.declaredSources?.forEach { source in
             let errors = validTargetPath(at: source)
             errors.forEach{ self.diags.emit(warning:: "Invalid: \(source): \(error)" }
         }
```